File: 130-surrounded-regions/surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        queue = deque()
        m = len(board[0])
        n = len(board)
        for c in range(m):
            if board[0][c] == 'O':
                queue.append((0,c))
            if board[n-1][c] == 'O':
                queue.append((n-1,c))
        for r in range(1,n-1):
            if board[r][0] == 'O':
                queue.append((r,0))
            if board[r][m-1] == 'O':
                queue.append((r,m-1))
        print(queue)
        def DFS(node):
            x,y = node[0],node[1]
            board[x][y] = -1
            for di,dj in [(1,0),(0,1),(-1,0),(0,-1)]:
                nr,nc = x + di, y + dj
                if 0 <= nr < n and 0 <= nc < m and board[nr][nc] == 'O':
                    board[nr][nc] = -1
                    DFS((nr,nc))
        while queue:
            px,py = queue.popleft()
            if board[px][py] == 'O':
                DFS((px,py))
        for r in range(n):
            for c in range(m):
                if board[r][c] == 'O':
                    board[r][c] = 'X'
                if board[r][c] == -1:
                    board[r][c] = 'O'

File: 130-surrounded-regions/surrounded-regions.java
class Solution {
    int [] x = new int[]{1,-1,0,0};
    int [] y = new int[]{0,0,1,-1};
    public void solve(char[][] board) {
        for(int i = 0;i<board.length;i++){
            for(int j = 0;j<board[i].length;j++){
                if((i == 0 || j == 0 || i == board.length-1 || j == board[0].length-1) && board[i][j] == 'O'){
                    fun(i,j,board);
                }
            }
        }
        for(int i = 0;i<board.length;i++){
            for(int j = 0;j<board[i].length;j++){
                if(board[i][j] == 'O'){
                    board[i][j] = 'X';
                }
            }
        }
        for(int i = 0;i<board.length;i++){
            for(int j = 0;j<board[i].length;j++){
                if(board[i][j] == 'Y'){
                    board[i][j] = 'O';
                }
            }
        }
    }
    public void fun(int i, int j, char [][] board){
        board[i][j] = 'Y';
        for(int k = 0;k<4;k++){
            int nexti = i + x[k];
            int nextj = j + y[k];
            if(nexti >= 0 && nexti < board.length && nextj >= 0 && nextj < board[0].length && board[nexti][nextj] == 'O'){
                fun(nexti,nextj,board);
            }
        }
    }
}